In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [32]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# Read salary data
file_path = Path("../Credit_risk/lending_data.csv")
df_lending = pd.read_csv(file_path)


# Review the DataFrame
df_lending.head()

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = df_lending["loan_status"]
# Separate the X variable, the features
X = df_lending.drop(columns="loan_status")

In [4]:
# Review the y variable Series (as dataframe)
pd.DataFrame(y)


loan_status
0                0
1                0
2                0
3                0
4                0
...            ...
77531            1
77532            1
77533            1
77534            1
77535            1

[77536 rows x 1 columns]

In [5]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
# sum of 0 and 1 should equal total number of rows (77536)
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(random_state=1)

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test})

Prediction  Actual
60914           0       0
36843           0       0
1966            0       0
70137           0       0
27237           0       0
...           ...     ...
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
#import accuracy_score
from sklearn.metrics import accuracy_score
# Print the balanced_accuracy score of the model
accuracy_score(y_test, predictions)

0.9918489475856377

In [27]:
#import confusion matrix
# Generate a confusion matrix for the model
#[[TN  FP]
#[FN  TP]]
from sklearn.metrics import confusion_matrix
cm1 = confusion_matrix(y_test, predictions)
cm1_df = pd.DataFrame(
    cm1, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
print("Confusion Matrix")
display(cm1_df)

Confusion Matrix


Predicted 0  Predicted 1
Actual 0        18663          102
Actual 1           56          563

In [12]:
# Print the classification report for the model
from sklearn.metrics import classification_report
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, predictions, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high-risk loan       0.85      0.91      0.88       619

      accuracy                           0.99     19384
     macro avg       0.92      0.95      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy of the model overall is 99.2% (the probability or predicting any input correctly). However, the model seems weaker when predicting the '1' or High-risk loans as the accuracy is only 88%. The accuracy of the model in predicting '0' or healthy loans is 100% and is based on 30x's more data to support it (which means it makes the regression model look very strong overall). 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
pip install -U imbalanced-learn


Note: you may need to restart the kernel to use updated packages.


In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# # Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# # Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

In [16]:
# Count the distinct values of the resampled labels data
unique_labels, counts = np.unique(y_resampled, return_counts=True)

# Display the counts for each unique label
for label, count in zip(unique_labels, counts):
    print(f'Label {label}: {count} instances')

Label 0: 56271 instances
Label 1: 56271 instances


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [17]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier2 = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
classifier2.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
predictions2 = classifier2.predict(X_test)
pd.DataFrame({"Prediction_oversample": predictions2, "Actual": y_test})

Prediction_oversample  Actual
60914                      0       0
36843                      0       0
1966                       0       0
70137                      0       0
27237                      0       0
...                      ...     ...
45639                      0       0
11301                      0       0
51614                      0       0
4598                       0       0
2793                       0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [18]:
# Print the balanced_accuracy score of the model 
accuracy_score(y_test, predictions2)

0.9938093272802311

In [28]:
# Generate a confusion matrix for the model
#[[TN  FP]
#[FN  TP]]

cm2 = confusion_matrix(y_test, predictions2)
cm2_df = pd.DataFrame(
    cm2, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
print("Confusion Matrix2")
#display(cm2_df)

Confusion Matrix


Predicted 0  Predicted 1
Actual 0        18649          116
Actual 1            4          615

In [29]:
#compare confusion matrix from original model vs over-sampled model
print("Confusion Matrix1 vs  Confusion Matrix2")
display(cm1_df, cm2_df)

Confusion Matrix1 vs  Confusion Matrix2


Predicted 0  Predicted 1
Actual 0        18663          102
Actual 1           56          563

Predicted 0  Predicted 1
Actual 0        18649          116
Actual 1            4          615

In [20]:
# Print the classification report for the model
target_names2 = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, predictions2, target_names=target_names2))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high-risk loan       0.84      0.99      0.91       619

      accuracy                           0.99     19384
     macro avg       0.92      0.99      0.95     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy of the model overall now 99.4%  as opposed to the previous model which was 99.2% (the probability or predicting any input correctly). The model with the oversampled data strengthened the model's ability to predict '1' or High-risk loans by 3%, as the model is now 91% accurate. The accuracy of the model in predicting '0' or healthy loans is the same as the previous model (100%). 

We should note that the Precision for the high-risk predictions is slightly lower in the over-sampled model (0.84 vs 0.85). Precision = (True Positives + False Positive)/True Positives. This means that the original model is better in catching false positives (loans marked as high-risk, but area actually healthy loans). 

Looking at the comparison of the two confusion matrices, it does look like False Positives are increased with the over-sampled model, but false-negatives are decreased. This means that loans that are actually high risk less likely to be predicted as healthy loans, which may be the most important factor in our case. 
